{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Age-association analyses of evolutionarily closest HA sequence pairs\n",
    "\n",
    "## Overview \n",
    "In this notebook, we demonstrate the analyses performed in our manuscript entitiled, \"Individual immune selection pressure has limited impact on seasonal influenza virus evolution\". Here, we show how we analysed for the hemagglutinin (HA) sequences of A/H3N2 collected between 2009 and July-2017.\n",
    "\n",
    "## Pre-analysis tasks \n",
    "### Sequence curation\n",
    "We collected all sequences from [GISAID](http://www.gisaid.org/) and [NCBI Genbank](http://www.ncbi.nlm.nih.gov/genbank/) databases that satisfy the following filters: \n",
    "  1. \\>90% of full HA nucleotide length \n",
    "  2. availability of patients' age data \n",
    "  3. availability of virus passage history that could be categorized as original clinical material, MDCK, SIAT and egg passage types (Typing of passage histories was based on [Chan et al., 2016](https://www.ncbi.nlm.nih.gov/pubmed/27604224)). \n",
    "\n",
    "In total, 10,514 HA nucleotide sequences were obtained for A/H3N2.\n",
    "\n",
    "We further processed the sequences obtained by: \n",
    "  1. Removed duplicated and low quality (\\>10% residues missing or ambiguous) were removed. \n",
    "  2. Sequences of the same virus but different passage histories were prioritized by clinical > MDCK > SIAT > egg and/or lower passage number. **At this point, there are 8,530 sequences for A/H3N2** (i.e. H3N2-HA-nuc.fa in [Files](https://github.com/alvinxhan/ageflu/tree/master/files) folder). \n",
    "  3. Clustered and removed identical nucleotide sequences using CD-HIT, leaving behind a representative strain for each cluster to yield a final set of non-redundant sequences. You can find the CD-HIT sequence cluster output file (e.g. H3N2-HA-nuc_cd-hit-non-redundant.clstr) in the [Files](https://github.com/alvinxhan/ageflu/tree/master/files) folder. \n",
    "  \n",
    "In total, there were 6,033 non-redudant sequences for A/H3N2. \n",
    "\n",
    "### Maximum-likelihood tree reconstruction\n",
    "Prior to the analyses presented here, we first reconstruct a maximum-likelihood (ML) phylogenetic tree using HA nucleotide sequences. Due to the large number of sequences (n=6,033 for A/H3N2) and the relatively low observed genetic divergence (overall mean nucleotide p-distance of A/H3N2 calculated by MEGA = 0.00964), conventional phylogenetic methods would be computationally expensive and practically infeasible. As such, we reconstructed our ML trees with RAxML and GARLI, using a nested inference approach that we have developed. More information on this phylogenetic inference methodology can be found in our paper. \n",
    "\n",
    "## Getting evolutionarily closest pairs\n",
    "We now can parse for evolutionarily closely-related virus pairs from the inferred phylogenetic tree (rooted, in NEWICK format). You can do so using [ageflu_getpairs.py](https://github.com/alvinxhan/ageflu/blob/master/scripts/ageflu_getpairs.py). \n",
    "\n",
    "* Additional inputs include: (i) FASTA alignment of all sequences (including identical nucleotide sequences), (ii) CD-HIT cluster file of non-redundant sequence clusters.\n",
    "* The age limits of children and adults can be changed using the '--child' and '--adult' arguments.\n",
    "* Note that the FASTA alignment containing reference sequences for HA-numbering ('H1pdm09_H3_FluB_NumberingRef.fa') must be in the same folder as the ageflu_getpairs.py script.\n",
    "* Tab-delimited output: ageflu_evol-closest-pairs.*.txt\n",
    "\n",
    "Here, we show the breakdown of the script:\n",
    "\n",
    "### Importing modules and defining functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from os.path import expanduser\n",
    "from decimal import *\n",
    "import re, sys, itertools\n",
    "import ete3\n",
    "import random\n",
    "\n",
    "random.seed(666)\n",
    "\n",
    "# parse fasta file\n",
    "def parsefasta(file, check_alg=1, check_dna=1, check_codon=1):\n",
    "    # check if file is of FASTA format\n",
    "    fhandle = filter(None, open(file, 'rU').readlines())\n",
    "    if not re.search('^>', fhandle[0]): sys.exit('\\nERROR: Incorrect sequence file format.\\n')\n",
    "\n",
    "    result = {}\n",
    "    for key, group in itertools.groupby(fhandle, lambda _: re.search('^>',_)):\n",
    "        if key:\n",
    "            header = group.next().strip().replace('>','')\n",
    "        else:\n",
    "            sequence = ''.join(map(lambda _:_.strip(),list(group)))\n",
    "            # dna sequences\n",
    "            if check_dna == 1 and set(list(sequence))&set(list('rhkdesqpvilmfyw')):\n",
    "                sys.exit('\\nERROR: Input must be DNA FASTA.\\n')\n",
    "            # codon-alignment\n",
    "            if check_codon == 1 and len(sequence)%3 > 0:\n",
    "                sys.exit('\\nERROR: DNA sequence file given must be codon-aligned.\\n')\n",
    "            result[header] = sequence\n",
    "    # alignment\n",
    "    if check_alg == 1 and len(set(map(len, result.values()))) != 1:\n",
    "        sys.exit('\\nERROR: Input sequence file must be an alignment.\\n')\n",
    "    return result\n",
    "\n",
    "# translate codon-aligned nucleotide sequences\n",
    "def translatedDNA(dna):\n",
    "    protein = []\n",
    "    for _ in xrange(0,len(dna),3):\n",
    "        codon = dna[_:_+3]\n",
    "        if codon in dnacodontable:\n",
    "            if dnacodontable[codon] == 'stop': break\n",
    "            protein.append(dnacodontable[codon])\n",
    "        elif re.match('---',codon): protein.append('-')\n",
    "        else: protein.append('X')\n",
    "    return ''.join(protein)\n",
    "\n",
    "# get pairwise substitutions\n",
    "def get_pairwise_substitutions(anc_seq, desc_seq):\n",
    "    # nucleotide sequences input\n",
    "    if set(list(anc_seq)) <= set(list('atgcn-')):\n",
    "        transitions, transversions, nonsyn_sub, syn_sub = 0, 0, 0, 0\n",
    "        for _ in xrange(0, len(anc_seq), 3):\n",
    "            anc_codon = anc_seq[_:_+3]\n",
    "            desc_codon = desc_seq[_:_+3]\n",
    "            if anc_codon == desc_codon:\n",
    "                continue\n",
    "            else:\n",
    "                unknown_res_binary = 0\n",
    "                nuc_sub = 0\n",
    "                for j in xrange(3):\n",
    "                    if re.match('(n|-)', anc_codon[j]) or re.match('(n|-)', desc_codon[j]):\n",
    "                        unknown_res_binary = 1\n",
    "                        continue\n",
    "                    elif anc_codon[j] != desc_codon[j]:\n",
    "                        if re.search('(ag|ga|ct|tc)', ''.join([anc_codon[j], desc_codon[j]])):\n",
    "                            transitions += 1\n",
    "                            nuc_sub += 1\n",
    "                        else:\n",
    "                            transversions += 1\n",
    "                            nuc_sub += 1\n",
    "\n",
    "                if unknown_res_binary == 1:\n",
    "                    continue\n",
    "                if dnacodontable[anc_codon] != dnacodontable[desc_codon]:\n",
    "                    nonsyn_sub += nuc_sub\n",
    "                else:\n",
    "                    syn_sub += nuc_sub\n",
    "        return {'transitions':transitions, 'transversions':transversions, 'nonsyn':nonsyn_sub, 'syn':syn_sub}\n",
    "    # protein sequences input\n",
    "    else:\n",
    "        mutation_list = []\n",
    "        for _ in xrange(len(anc_seq)):\n",
    "            if re.search('(-|X)', anc_seq[_]) or re.search('(-|X)', desc_seq[_]):\n",
    "                continue\n",
    "            elif anc_seq[_] != desc_seq[_]:\n",
    "                mutation_list.append('{}{}{}'.format(anc_seq[_], _+1, desc_seq[_]))\n",
    "        return mutation_list\n",
    "\n",
    "def get_least_unknown_residue_sequences(seq_dict, dist_dict_to_edit={}):\n",
    "    seqid_to_unknownrescount = {seqid:len(re.findall('(-|n|X)', sequence, re.I)) for seqid, sequence in seq_dict.items()}\n",
    "    seqid_with_least_unknown_res = [seqid for seqid, count in seqid_to_unknownrescount.items() if count == min(seqid_to_unknownrescount.values())]\n",
    "    if len(dist_dict_to_edit) > 0:\n",
    "        for seqid in list(set(seq_dict.keys())-set(seqid_with_least_unknown_res)):\n",
    "            del dist_dict_to_edit[seqid]\n",
    "        return seqid_with_least_unknown_res, dist_dict_to_edit\n",
    "    else:\n",
    "        return seqid_with_least_unknown_res\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Define inputs \n",
    "This is defined via command line in the [ageflu_getpairs.py](https://github.com/alvinxhan/ageflu/blob/master/scripts/ageflu_getpairs.py) script. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "class params:\n",
    "    tree = './files/H3N2-HA-nuc.rooted.nwk' # Phylogenetic tree file in NEWICK format \n",
    "    aln = './files/H3N2-HA-nuc.fa' # Nucleotide alignment of HA sequences (pre-CD-HIT)\n",
    "    nr = './files/H3N2-HA-nuc_cd-hit-non-redundant.clstr' # CD-HIT identical sequence cluster file \n",
    "    child = [0, 5] # age range of children \n",
    "    adult = [35, 120] # age range of adults \n",
    "    maxmut = 5 # maxmimum number of amino acid substitutions allowable in an evolutionarily closest pair \n",
    "    patdist = 0.007 # maximum patristic distance between sequences in a closest pair\n",
    "    order = 3 # max ordinal of pairs "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parse FASTA file and sort CD-HIT identical sequence clusters into age categories (child or adult) based on input children and adults age ranges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EXAMPLE of the age category of a CD-HIT cluster:\n",
      "Representative sequence = A/SouthCarolina/22/2016_HA_EPIISL225062_2016.39767283_ORI/CLI_A/H3N2_LAT33.836081LON-81.163724_AGE18\n",
      "Age category: U\n",
      "Member sequences:\n",
      "A/SouthCarolina/19/2016_HA_EPIISL225061_2016.39493498_ORI/CLI_A/H3N2_LAT33.836081LON-81.163724_AGE19\n",
      "A/SouthCarolina/25/2016_HA_EPIISL225067_2016.39767283_ORI/CLI_A/H3N2_LAT33.836081LON-81.163724_AGE19\n",
      "A/SouthCarolina/17/2016_HA_EPIISL225128_2016.39219713_ORI/CLI_A/H3N2_LAT33.836081LON-81.163724_AGE25\n",
      "A/SouthCarolina/22/2016_HA_EPIISL225062_2016.39767283_ORI/CLI_A/H3N2_LAT33.836081LON-81.163724_AGE18\n"
     ]
    }
   ],
   "source": [
    "# query subtype analysed \n",
    "try:\n",
    "    query_subtype = re.search('(H3N2|pH1N1|H1N1pdm09|BVic|BYam)', params.tree).group().upper()\n",
    "    if query_subtype == 'PH1N1':\n",
    "        query_subtype = 'H1N1PDM09'\n",
    "except:\n",
    "    query_subtype = raw_input('\\nWARNING: Can\\'t parsed query subtype from tree name. Enter subtype (H3N2/pH1N1/BVic/BYam): ')\n",
    "\n",
    "# output filename\n",
    "outfname = '{}C{}_{}A{}_PD{}_MM{}_CL{}_{}'.format(params.child[0], params.child[-1], params.adult[0], params.adult[-1], params.patdist, params.maxmut, params.order, re.sub('[^/]*/', '', params.tree))\n",
    "\n",
    "# parse fasta alignment \n",
    "dnacodontable = {'ttt':'F', 'ttc':'F', 'tta':'L', 'ttg':'L', 'ctt':'L', 'ctc':'L', 'cta':'L', 'ctg':'L', 'att':'I', 'atc':'I', 'ata':'I', 'atg':'M', 'gtt':'V', 'gtc':'V', 'gta':'V', 'gtg':'V', 'tct':'S', 'tcc':'S', 'tca':'S', 'tcg':'S', 'cct':'P', 'ccc':'P', 'cca':'P', 'ccg':'P', 'act':'T', 'acc':'T', 'aca':'T', 'acg':'T', 'gct':'A', 'gcc':'A', 'gca':'A', 'gcg':'A', 'tat':'Y', 'tac':'Y', 'taa':'stop', 'tag':'stop', 'cat':'H', 'cac':'H', 'caa':'Q', 'cag':'Q', 'aat':'N', 'aac':'N', 'aaa':'K', 'aag':'K', 'gat':'D', 'gac':'D', 'gaa':'E', 'gag':'E', 'tgt':'C', 'tgc':'C', 'tga':'stop', 'tgg':'W', 'cgt':'R', 'cgc':'R', 'cga':'R', 'cgg':'R', 'agt':'S', 'agc':'S', 'aga':'R', 'agg':'R', 'ggt':'G', 'ggc':'G', 'gga':'G', 'ggg':'G'}\n",
    "\n",
    "fdat_nuc = parsefasta(params.aln)\n",
    "fdat_aa = {k:translatedDNA(v) for k,v in fdat_nuc.items()}\n",
    "isolateid_to_fdatheader = {re.search('(GBISL|EPIISL|GB_ISL_|EPI_ISL_)\\d+', header).group().replace('_', ''):header for header in fdat_nuc.keys()}\n",
    "fdatheader_to_isolateid = {v:k for k,v in isolateid_to_fdatheader.items()}\n",
    "isolateid_to_age = {isolateid:int(re.search('AGE(\\d+)', header).group(1)) for isolateid, header in isolateid_to_fdatheader.items()}\n",
    "\n",
    "# sequences to ignore (outside of child_min and adult_max age)\n",
    "isolates_to_ignore = [isolateid for isolateid, age in isolateid_to_age.items() if age < params.child[0] or age > params.adult[-1]]\n",
    "\n",
    "# parse clstr file and sort each identical sequence cluster (>1 member sequence) into children/adult age categories \n",
    "# isolates_to_ignore now include those of ambiguous age categories (between child_max and adult_min) \n",
    "# + 0 pat dist sequences with > minimal number of n/-\n",
    "isolateid_to_representatives, isolateid_to_agecategory = {}, {} \n",
    "fhandle = filter(None, open(params.nr, 'rU').readlines())\n",
    "for key, group in itertools.groupby(fhandle, lambda _: re.search('^>Cluster', _)):\n",
    "    if not key:\n",
    "        cluster = list(group)\n",
    "        if len(cluster) > 1:\n",
    "            cluster = [re.search('(GBISL|EPIISL|GB_ISL_|EPI_ISL_)\\d+', header).group().replace('_', '') for header in cluster]\n",
    "            cluster_age = [isolateid_to_age[isolateid] for isolateid in cluster]\n",
    "            # all children or adults sequence clusters\n",
    "            if all([params.child[0] <= age <= params.child[-1] for age in cluster_age]):\n",
    "                isolateid_to_agecategory[isolateid] = 'C'\n",
    "                for isolateid in cluster:\n",
    "                    isolateid_to_representatives[isolateid] = cluster\n",
    "            elif all([params.adult[0] <= age <= params.adult[-1] for age in cluster_age]):\n",
    "                isolateid_to_agecategory[isolateid] = 'A'\n",
    "                for isolateid in cluster:\n",
    "                    isolateid_to_representatives[isolateid] = cluster\n",
    "            # uncategorized age present in sequence cluster \n",
    "            elif all([params.child[0] <= age <= params.adult[-1] for age in cluster_age]) and any([params.child[-1] < age < params.adult[0] for age in cluster_age]):\n",
    "                isolateid_to_agecategory[isolateid] = 'U'\n",
    "                for isolateid in cluster:\n",
    "                    isolateid_to_representatives[isolateid] = [member for _, member in enumerate(cluster) if params.child[-1] < cluster_age[_] < params.adult[0]]\n",
    "            else:\n",
    "                isolates_to_ignore = list(set(isolates_to_ignore)|set(cluster))\n",
    "\n",
    "# print example\n",
    "example_isolateid = random.choice(isolateid_to_agecategory.keys())\n",
    "print ('EXAMPLE of the age category of a CD-HIT cluster:\\nRepresentative sequence = {}'.format(isolateid_to_fdatheader[example_isolateid]))\n",
    "print ('Age category: {}'.format(isolateid_to_agecategory[example_isolateid]))\n",
    "print ('Member sequences:\\n{}'.format('\\n'.join([isolateid_to_fdatheader[_] for _ in isolateid_to_representatives[example_isolateid]])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parsing phylogenetic tree using ete3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "    try:\n",
    "        tree = ete3.Tree(params.tree)\n",
    "    except:\n",
    "        sys.exit('\\nERROR: Unable to parse tree using ete3.\\n')\n",
    "    tree.ladderize()\n",
    "    # get nodes to isolate ids\n",
    "    leaf_to_isolateid = {leaf:re.search('(EPI_ISL_|EPIISL|GB_ISL_|GBISL)\\d+', leaf.get_leaf_names()[0]).group().replace('_', '') for leaf in tree.get_leaves()}\n",
    "    isolateid_to_leaf = {v:k for k,v in leaf_to_isolateid.items()}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
